### DISCUSSION TITLE
Lesson 12: Why did we make the sender in the `testRevertsIfCollateralZero()` test as DSCEngine instead of a prank user?

### DISCUSSION THREAD

**Discussion by alfheimrShiven** [Cyfrin/foundry-full-course-cu#790] [Q&A]
So in the following test you will see @PatrickAlphaC has made the collateral sender as `DSCEngine` using 
`ERC20Mock(weth).approve(address(dsce), amountCollateral)` and then called the `dsce.depositCollateral(weth, amountCollateral)` function. Why was this done? Why didn't we just send the collateral from the prank user like follows?

```solidity
function testRevertsIfCollateralZero() public {
        vm.startPrank(user);
        // ERC20Mock(weth).approve(address(dsce), amountCollateral);

        vm.expectRevert(DSCEngine.DSCEngine__NeedsMoreThanZero.selector);
        dsce.depositCollateral(weth, 0);
        vm.stopPrank();
    }
```

Generally, any random user should be able to deposit collateral right?

**Comment by theirrationalone** [comment]
`ERC20` protocol specific. 

According to the protocol, Whenever we need to send / receive, crypto asset, We first need to approve to someone to let him/her/it, use our crypto asset. Since we're the deployer/owner who deployed protocol, we have choice/right, to do whatever we want from our protocol using facilitating features provided by protocol.

Therefore, You can skip that approval part by implementing a logic where everybody has a approval Or few ones who have approval, for your dsc, if you want to. Although that would be considered vulnerable. 

Sorry if i'm wrong here!

**Reply by alfheimrShiven** [reply]
Sorry couldn't understand v well.

**Reply by theirrationalone** [reply]
Lemme try to explain with an example.... 

If someone wants to take some money from Government's treasury. First the beneficiary(that someone) needs to have some kind of approval paper with a consent & signature of the Governor. Let's say, the beneficiary got the governor's approval with all consent & signature then we needs to submit that approval to treasury keeping department.

Treasury keeping department shall record an entry with the provided approval & beneficiary's details, and shall give the amount mentioned into the approval to the beneficiary. 

Hope, Now you can understand protocol's working. 😅
